src/main/sensors/opticalflow.c

```diff
@@ -0,0 +1,193 @@
+#include "common/utils.h"
+#include "common/filter.h"
+
+#include "config/config.h"
+#include "config/feature.h"
+
+#include "fc/runtime_config.h"
+
+#include "scheduler/scheduler.h"
+
+#include "pg/pg.h"
+#include "pg/pg_ids.h"
+
+#include "drivers/time.h"
+#include "drivers/rangefinder/rangefinder.h"
+#include "drivers/rangefinder/rangefinder_lidarmt.h"
+
+#include "io/beeper.h"
+
+#include "sensors/sensors.h"
+#include "sensors/gyro.h"
+#include "sensors/acceleration.h"
+#include "sensors/opticalflow.h"
+
+#define OPTICALFLOW_CALIBRATION_DURATION_MS 30000
+#define RATE_SCALE_RESOLUTION (1000.0f)
+
+// static prototypes
+static void applySensorRotation(vector2_t * dst, vector2_t * src);
+static void applyLPF(vector2_t * flowRates);
+
+PG_REGISTER_WITH_RESET_TEMPLATE(opticalflowConfig_t, opticalflowConfig, PG_OPTICALFLOW_CONFIG, 0);
+
+PG_RESET_TEMPLATE(opticalflowConfig_t, opticalflowConfig,
+    .opticalflow_hardware = OPTICALFLOW_NONE,
+    .rotation = 0,
+    .flow_lpf = 0,
+    .flip_y = 0
+);
+
+static opticalflow_t opticalflow;
+float cosRotAngle = 1.0f;
+float sinRotAngle = 0.0f;
+static pt2Filter_t xFlowLpf, yFlowLpf;
+
+// ======================================================================
+// =================== Opticalflow Main Functions =======================
+// ======================================================================
+static bool opticalflowDetect(opticalflowDev_t * dev, uint8_t opticalflowHardwareToUse) {
+    UNUSED(dev);
+
+    opticalflowType_e opticalflowHardware = OPTICALFLOW_NONE;
+    requestedSensors[SENSOR_INDEX_OPTICALFLOW] = opticalflowHardwareToUse;
+
+    switch (opticalflowHardwareToUse) {
+        case OPTICALFLOW_MT:
+#ifdef USE_RANGEFINDER_MT
+            if (mtOpticalflowDetect(dev, rangefinderConfig()->rangefinder_hardware)) {
+                opticalflowHardware = OPTICALFLOW_MT;
+                rescheduleTask(TASK_OPTICALFLOW, TASK_PERIOD_MS(dev->delayMs));
+            }
+#endif
+            break;
+
+        case OPTICALFLOW_NONE:
+            opticalflowHardware = OPTICALFLOW_NONE;
+            break;
+    }
+
+    if (opticalflowHardware == OPTICALFLOW_NONE) {
+        sensorsClear(SENSOR_OPTICALFLOW);
+        return false;
+    }
+
+    detectedSensors[SENSOR_INDEX_OPTICALFLOW] = opticalflowHardware;
+    sensorsSet(SENSOR_OPTICALFLOW);
+    return true;    
+}
+
+bool opticalflowInit(void) {
+    if (!opticalflowDetect(&opticalflow.dev, opticalflowConfig()->opticalflow_hardware)) {
+        return false;
+    }
+
+    opticalflow.dev.init(&opticalflow.dev);
+    opticalflow.quality = OPTICALFLOW_NO_NEW_DATA;
+    opticalflow.rawFlowRates.x = 0;
+    opticalflow.rawFlowRates.y = 0;
+    opticalflow.processedFlowRates.x = 0;
+    opticalflow.processedFlowRates.y = 0;
+    opticalflow.timeStampUs = micros();
+
+        cosRotAngle = cosf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
+        sinRotAngle = sinf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
```

Comment:
Suggestion did not include spacing :)
```suggestion
    cosRotAngle = cosf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
    sinRotAngle = sinf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
```